### PR TITLE
Buffer and spill only required channels for distinct aggregates

### PR DESCRIPTION
**Pull request by pgupta2**
Distinct aggregate spilling logic buffers the input pages and spill them
when memoryRevoke is called. Currently, entire input page is buffered and
spilled. The aggregate is not going to consume all the data in the page.
We can buffer and spill only the required channels from the input page.
This reduces the memory required for buffering data and also reduces bytes
that needs to be spilled.

Test plan - Unit Tests.  Also, Performed verifier runs for DISTINCT queries to check for correctness.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java [hunk 1]
```diff
@@ -791,6 +823,11 @@ public void prepareFinal()
             delegate.prepareFinal();
         }
 
+        private static RunLengthEncodedBlock createNullBlock(Type type, int positionCount)
+        {
+            return new RunLengthEncodedBlock(type.createBlockBuilder(null, 1).appendNull().build(), positionCount);
+        }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:829]: We can inline this function as it is only used by one place. Also, let's use `RunLengthEncodedBlock.create` which is simpler then `new RunLengthEncodedBlock`.

**Review by highker — APPROVED** [review]
This is an interesting fix :D

**Comment by highker** [comment]
Curious for the verifier run: have we used a flash-device cluster and checked there was spilling happening?

**Comment by pgupta2** [comment]
> Curious for the verifier run: have we used a flash-device cluster and checked there was spilling happening?

For verifier runs, I picked up DISTINCT queries and basically ran control version on presto-classic/presto-on-spark with spilling disabled and test version on Presto-on-Spark with spilling enabled for correctness testing. I have not done any cluster specific testing for this. If this is something we should test, can you point out the step for doing this testing?

**Comment by highker** [comment]
@varungajjala, do we know what would be a recommended way to test spilling on flash devices?

**Pull request merged**